src/librustc_codegen_llvm/va_arg.rs

```diff
@@ -0,0 +1,141 @@
+
+use builder::Builder;
+use common::C_i32;
+use mir::operand::OperandRef;
+use rustc::ty::layout::{Align, HasDataLayout, LayoutOf, Size};
+use rustc::ty::Ty;
+use type_::Type;
+use type_of::LayoutLlvmExt;
+use value::Value;
+
+#[allow(dead_code)]
+fn round_pointer_up_to_alignment(
+    bx: &Builder<'a, 'll, 'tcx>,
+    addr: &'ll Value,
+    align: Align,
+    ptr_ty: &'ll Type
+) -> &'ll Value {
+    let mut ptr_as_int = bx.ptrtoint(addr, Type::isize(bx.cx));
+    ptr_as_int = bx.add(ptr_as_int, C_i32(bx.cx, (align.abi() as i32) - 1));
+    ptr_as_int = bx.and(ptr_as_int, C_i32(bx.cx, -(align.abi() as i32)));
+    bx.inttoptr(ptr_as_int, ptr_ty)
+}
+
+fn emit_direct_ptr_va_arg(
+    bx: &Builder<'a, 'll, 'tcx>,
+    list: OperandRef<'ll, 'tcx>,
+    llty: &'ll Type,
+    size: Size,
+    align: Align,
+    slot_size: Align,
+    allow_higher_align: bool
+) -> (&'ll Value, Align) {
+    let va_list_addr = if list.layout.llvm_type(bx.cx) != Type::i8p(bx.cx).ptr_to() {
+        bx.bitcast(list.immediate(), Type::i8p(bx.cx).ptr_to())
+    } else {
+        list.immediate()
+    };
+
+    let ptr = bx.load(va_list_addr, bx.tcx().data_layout.pointer_align);
+
+    let (addr, addr_align) = if allow_higher_align && align > slot_size {
+        (round_pointer_up_to_alignment(bx, ptr, align, Type::i8p(bx.cx)), align)
+    } else {
+        (ptr, slot_size)
+    };
+
+
+    let aligned_size = size.abi_align(slot_size).bytes() as i32;
+    let full_direct_size = C_i32(bx.cx, aligned_size);
+    let next = bx.inbounds_gep(addr, &[full_direct_size]);
+    bx.store(next, va_list_addr, bx.tcx().data_layout.pointer_align);
+
+    if size.bytes() < slot_size.abi() &&
+            &*bx.tcx().sess.target.target.target_endian == "big" {
+        let adjusted_size = C_i32(bx.cx, (slot_size.abi() - size.bytes()) as i32);
+        let adjusted = bx.inbounds_gep(addr, &[adjusted_size]);
+        (bx.bitcast(adjusted, llty.ptr_to()), addr_align)
+    } else {
+        (bx.bitcast(addr, llty.ptr_to()), addr_align)
+    }
+}
+
+fn emit_ptr_va_arg(
+    bx: &Builder<'a, 'll, 'tcx>,
+    list: OperandRef<'ll, 'tcx>,
+    target_ty: Ty<'tcx>,
+    indirect: bool,
+    slot_size: Align,
+    allow_higher_align: bool
+) -> &'ll Value {
+    let layout = bx.cx.layout_of(target_ty);
+    let (llty, size, align) = if indirect {
+        (bx.cx.layout_of(bx.cx.tcx.mk_imm_ptr(target_ty)).llvm_type(bx.cx),
+         bx.cx.data_layout().pointer_size,
+         bx.cx.data_layout().pointer_align)
+    } else {
+        (layout.llvm_type(bx.cx),
+         layout.size,
+         layout.align)
+    };
+    let (addr, addr_align) = emit_direct_ptr_va_arg(bx, list, llty, size, align,
+                                                    slot_size, allow_higher_align);
+    if indirect {
+        let tmp_ret = bx.load(addr, addr_align);
+        bx.load(tmp_ret, align)
+    } else {
+        bx.load(addr, addr_align)
+    }
+}
+
+pub(super) fn emit_va_arg(
```

Comment:
Much of what exists in this file is my attempt at translating `emitVoidPtrVAArg` in clang (found in `lib/CodeGen/TargetInfo.cpp`) to rust.